enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,476 @@
+  schema: patchy
+- name: InputBulldog
+  application:
+    namespaces: [ jill, jane ]
+  structured: json
+  schema: bulldog
+pipeline:
+- inputRefs: [ InputPatchy, InputBulldog ]
+  outputRefs: default
+```
+
+##### Example 2: Identify applications by labels.
+
+The applications using each schema must be identified by k8s labels.
+Note this is often _already done_ as part of a Deployment.
+Deployments usually label pods by the application they run, which determines their log format.
+
+**Note**: the label `selector` below is defined by the [forwarder label selectors proposal][clf-labels]
+
+```yaml
+inputs:
+- name: InputPatchy
+  applications:
+    selector:
+      matchLabels: { app: patchy }
+  structured: json
+  schema: patchy
+- name: InputBulldog
+  applications:
+    selector:
+      matchLabels: { app: bulldog }
+  structured: json
+  schema: bulldog
+pipelines:
+- inputRefs: [ InputPatchy, InputBulldog ]
+  outputRefs: default
+```
+
+#### I want to forward structured and unstructured records to the same output
+
+Structured records will have the `structured` field, non-structured will have the `message` field.
+
+```yaml
+inputs:
+- name: InputMaybeJSON
+  application: {}
+  structured: json
+outputs:
+- name: OutputMaybeJSON
+  type: fluentdForward
+  ...
+pipelines:
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputMaybeJSON ]
+```
+
+#### I want a replacement for the defunct MERGE_JSON_LOG feature
+
+```yaml
+inputs:
+- name: InputMaybeJSON
+  application: {}
+  structured: json
+outputs:
+- name: OutputMerged
+  type: fluentdForward
+  ...
+pipelines:
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputMerged ]
+  modify:
+    merge: true
+```
+
+#### I want to  forward structured and unstructured logs from the same input to different outputs
+
+```yaml
+inputs:
+- name: InputMaybeJSON
+  application: {}
+  structured: json
+outputs:
+- name: OutputJSON
+  type: fluentdForward
+  ...
+outputs:
+- name: OutputNotJSON
+  type: fluentdForward
+  ...
+pipelines:
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputJSON ]
+  filter:
+    require: [structured]
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputNotJSON ]
+  filter:
+    reject: [structured]
+```
+
+#### I want to forward structured records to one output, and all records in unstructured form to another
+
+```yaml
+inputs:
+- name: InputMaybeJSON
+  application: {}
+  structured: json
+outputs:
+- name: OutputJSON
+  type: fluentdForward
+  ...
+outputs:
+- name: OutputAllText
+  type: fluentdForward
+  ...
+pipelines:
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputJSON ]
+  filter:
+    require: [structured]
+- inputRefs: [ InputMaybeJSON ]
+  outputRefs: [ OutputAllText ]
+  modify:
+    remove: [structured] # Replaced with an unstructured `message` field.
+```
+
+### Implementation Details
+
+#### Elasticsearch version-specific workarounds.
+
+The existing Elasticsearch logging store requires "flattening" of the `kubernetes.labels` map:
```

Comment:
I don't think that its required... but there wasn't a good method for querying json objects before so we flattened them

Reply:
I'd like to understand that better.
I want the "default" presentation of labels to be the original map, most JSON consumers will want that.
I believe that ES pipelines we can also handle the map form fine, since the pipeline knows about the data model and can handle kubernetes.labels in a way that works.
I want to make sure we remain compatible with existing ES to handle the transition, so if necessary we can keep flattening in that specific case at the forwarder - but if not necessary then we can just forward labeles as maps everywhere.